### PR TITLE
Add compression-webpack-plugin

### PR DESCRIPTION
**Pull request by tbrisker**
This depends on https://github.com/theforeman/foreman/pull/3883 being merged.

For plugin updates, please indicate which repos this should be built into:

* [x] Nightly
* [ ] 1.14
* [ ] 1.13
* [ ] 1.12

See Foreman's [plugin maintainer documentation](http://projects.theforeman.org/projects/foreman/wiki/How_to_Create_a_Plugin#Release-strategies) for more information.

---

### CODE REVIEW

**Comment by lzap** [comment]
FYI core was merged.

**Comment by domcleal** [comment]
This is still failing to build, please fix the failure and ensure comments/FIXME comments are resolved in the spec file.

**Comment by tbrisker** [comment]
@domcleal updated the spec file, hopefully tests will pass now, thanks.

**Comment by tbrisker** [comment]
@domcleal fixed the failing build, seems to work now.

**Comment by tbrisker** [comment]
looks like this is pulling in a ton of unwanted optional dependencies due to an issue with how npm registry lists dependcies. https://github.com/npm/npm-remote-ls/pull/32 should fix it and significantly reduce the size of this package, I'll try to get this fixed.

**Comment by ehelms** [comment]
@tbrisker is the npm-remote-ls fix required to generate the package or to build the package? If the former, could we build out the spec and dependencies by hand so as to get this built given it will be required to get nightly builds going.

**Comment by dLobatog** [comment]
@tbrisker I think this could be built manually as 3 packages with the single strategy, it'd save us time until they merge the npm-remote-ls fix.

@ehelms It's not strictly required but without it, the bundle package will be huge, when there are only 3 mandatory dependencies in fact.

**Comment by tbrisker** [comment]
Pushed again with only the required deps using the single strategy, @dLobatog @domcleal  please re-review.

**Comment by dLobatog** [comment]
@tbrisker This is missing adding the package to `foreman.spec`

File: comps/comps-foreman-fedora24.xml [hunk 1]
```diff
@@ -29,13 +29,15 @@
       <packagereq type="default">foreman-selinux</packagereq>
       <packagereq type="default">foreman-sqlite</packagereq>
       <packagereq type="default">foreman-vmware</packagereq>
+      <packagereq type="default">nodejs-async</packagereq>
```
Comment (dLobatog) [inline comps/comps-foreman-fedora24.xml]: Run `./comps_doc.sh`, there are no -doc packages in either of the comps

File: nodejs-async/nodejs-async.spec [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+%if 0%{?enable_tests}
+%check
+%{nodejs_symlink_deps} --check
+%endif
+
+%files
+%{nodejs_sitelib}/%{npm_name}
+
+%doc LICENSE
```
Comment (dLobatog) [inline nodejs-async/nodejs-async.spec]:
```
%files
%doc LICENSE
....

%files doc
%doc README.md
```

preferred, one files section for docs and another for regular files, license and binaries

File: nodejs-compression-webpack-plugin/nodejs-compression-webpack-plugin.spec [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+%if 0%{?enable_tests}
+%check
+%{nodejs_symlink_deps} --check
+%endif
+
+%files
+%{nodejs_sitelib}/%{npm_name}
+
+%doc README.md
```
Comment (dLobatog) [inline nodejs-compression-webpack-plugin/nodejs-compression-webpack-plugin.spec]: Same as the other spec, have another `%files doc` section for this

File: nodejs-webpack-sources/nodejs-webpack-sources.spec [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+%if 0%{?enable_tests}
+%check
+%{nodejs_symlink_deps} --check
+%endif
+
+%files
+%{nodejs_sitelib}/%{npm_name}
+
+%doc README.md
```
Comment (dLobatog) [inline nodejs-webpack-sources/nodejs-webpack-sources.spec]: Same as the other spec, have another `%files doc` for this.

File: nodejs-webpack-sources/nodejs-webpack-sources.spec [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
+%prep
+%setup -q -n package
+
+%build
+%nodejs_symlink_deps --build
+
+%install
+mkdir -p %{buildroot}%{nodejs_sitelib}/%{npm_name}
+cp -pfr .editorconfig .eslintrc .gitattributes .jsbeautifyrc .npmignore .travis.yml README.md lib npm-debug.log.18166252 package.json test %{buildroot}%{nodejs_sitelib}/%{npm_name}
```
Comment (dLobatog) [inline nodejs-webpack-sources/nodejs-webpack-sources.spec]: There seems to be a bunch of cruft in here that we don't really need to package, remove at your own judgement

File: nodejs-compression-webpack-plugin/nodejs-compression-webpack-plugin.spec [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+
+%prep
+%setup -q -n package
+
+%build
+%nodejs_symlink_deps --build
+
+%install
+mkdir -p %{buildroot}%{nodejs_sitelib}/%{npm_name}
+cp -pfr .gitattributes .npmignore README.md index.js package.json %{buildroot}%{nodejs_sitelib}/%{npm_name}
```
Comment (dLobatog) [inline nodejs-compression-webpack-plugin/nodejs-compression-webpack-plugin.spec]: I don't think we need these dotfiles

**Comment by dLobatog** [comment]
@tbrisker Nearly there - I think it'd be 100% ready after the changes I've mentioned in the review. Thanks!

**Comment by tbrisker** [comment]
@dLobatog hopefully all good now

**Comment by dLobatog** [comment]
👍  from me provided tests pass

File: foreman/foreman.spec [hunk 7]
```diff
@@ -219,6 +219,8 @@ BuildRequires: npm(babel-preset-react) < 7.0.0
 BuildRequires: npm(babel-preset-react) >= 6.5.0
 BuildRequires: npm(c3) >= 0.4.11
 BuildRequires: npm(c3) < 1.0.0
+BuildRequires: npm(compression-webpack-plugin) >= 0.3.2
```
Comment (dLobatog) [inline foreman/foreman.spec]: Looks like these changes have to go on a different PR to ensure this gets built after the deps.

**Comment by dLobatog** [comment]
@domcleal Anything missing here? Seems like this PR is required for nightlies to get going

**Comment by domcleal** [comment]
webpack-sources appears to depend on source-map and source-list-map, but these dependencies aren't either here or bundled. Can you check the package deps resolve?

**Comment by tbrisker** [comment]
@domcleal Thanks, updated with those 2 packages as well.

File: nodejs-source-list-map/nodejs-source-list-map.spec [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+
+%install
+mkdir -p %{buildroot}%{nodejs_sitelib}/%{npm_name}
+cp -pfr README.md lib package.json %{buildroot}%{nodejs_sitelib}/%{npm_name}
+%nodejs_symlink_deps
+
+%if 0%{?enable_tests}
+%check
+%{nodejs_symlink_deps} --check
+#$CHECK
```
Comment (domcleal) [inline nodejs-source-list-map/nodejs-source-list-map.spec]: Useless comment.

File: nodejs-source-map/nodejs-source-map.spec [hunk 9]
```diff
@@ -0,0 +1,61 @@
+%global npm_name source-map
+%global enable_tests 1
+
+%{?nodejs_find_provides_and_requires}
+
+Name: nodejs-%{npm_name}
+Version: 0.5.6
+Release: 1%{?dist}
+Summary: Generates and consumes source maps
+License: BSD-3-Clause
```
Comment (domcleal) [inline nodejs-source-map/nodejs-source-map.spec]: BSD

File: nodejs-source-map/nodejs-source-map.spec [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+
+%install
+mkdir -p %{buildroot}%{nodejs_sitelib}/%{npm_name}
+cp -pfr CHANGELOG.md LICENSE README.md dist lib package.json source-map.js %{buildroot}%{nodejs_sitelib}/%{npm_name}
+%nodejs_symlink_deps
+
+%if 0%{?enable_tests}
+%check
+%{nodejs_symlink_deps} --check
+#$CHECK
```
Comment (domcleal) [inline nodejs-source-map/nodejs-source-map.spec]: remove

File: nodejs-source-map/nodejs-source-map.spec [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+%endif
+
+%files
+%{nodejs_sitelib}/%{npm_name}
+%doc LICENSE
+%doc CHANGELOG.md
```
Comment (domcleal) [inline nodejs-source-map/nodejs-source-map.spec]: -doc subpackage

File: comps/comps-foreman-fedora24.xml (hunk 1 above)
Comment (domcleal) [inline comps/comps-foreman-fedora24.xml]: nodejs-async 1.5's already in the distribution, is the older dep going to resolve correctly, since it's `0.2.x`? Bundling the dependency would be preferable I think in this case.

**Comment by tbrisker** [comment]
Yet another try, back to bundled.

**Comment by domcleal** [comment]
Please check the RPM builds, there is a failure.

**Comment by tbrisker** [comment]
I give up. The time and effort wasted on making these useless rpms is ridiculous, instead of just running an `npm install` and adding node_modules to the sources passed to koji prior to build.
Anyone else who want to get this building is welcome to, I'm not going to sink any more hours into figuring out why this fails to build.
